hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java

```diff
@@ -1118,15 +1118,28 @@ public AbfsInputStream openFileForRead(final Path path,
 
       String relativePath = getRelativePath(path);
 
-      final AbfsRestOperation op = client
-          .getPathStatus(relativePath, false, tracingContext);
+      AbfsRestOperation op;
+      if (OperativeEndpoint.isReadEnabledOnDFS(getPrefixMode(), abfsConfiguration)) {
+        op = client.getBlobProperty(new Path(relativePath), tracingContext);
+      } else {
+        op = client
+                .getPathStatus(relativePath, false, tracingContext);
+      }
+
       perfInfo.registerResult(op.getResult());
 
-      final String resourceType = op.getResult().getResponseHeader(HttpHeaderConfigurations.X_MS_RESOURCE_TYPE);
+      boolean isDirectory;
+      if (getPrefixMode() == PrefixMode.BLOB) {
+        isDirectory = Boolean.parseBoolean(op.getResult().getResponseHeader(X_MS_META_HDI_ISFOLDER));
+      } else {
+        final String resourceType = op.getResult().getResponseHeader(HttpHeaderConfigurations.X_MS_RESOURCE_TYPE);
+        isDirectory = parseIsDirectory(resourceType);
+      }
+
       final long contentLength = Long.parseLong(op.getResult().getResponseHeader(HttpHeaderConfigurations.CONTENT_LENGTH));
       final String eTag = op.getResult().getResponseHeader(HttpHeaderConfigurations.ETAG);
 
-      if (parseIsDirectory(resourceType)) {
+      if (isDirectory) {
```

Comment:
Why we need to check we have to read on dfs or blob here?
What we need to check here is that are the blob properties which we will get from the getFileStatus.

In my opinion, we should not change this piece of code. One thing though, direct call of `client.getPathStatus()` is bad from here. It should have called the store's `getFileStatus`. Lets change it to call the store's method. Benefit of that is: when the getPathStatus API Pr will come, it would have no change in `openFileForRead`. And for `openFileForRead`, `getFileStatus` is abstract and it expects that it gives the filestatus of the asked path.